### DISCUSSION TITLE
Warning: 'python-mutagen' formula is unreadable: No available formula with the name "python-mutagen". Did you mean python-mako?

### DISCUSSION THREAD

**Discussion by chrisfinazzo** [Homebrew/discussions#5179] [Everyday usage] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.11-36-gd55fa09
ORIGIN: https://github.com/Homebrew/brew
HEAD: d55fa09d1ee877859dd22f4f8129f7fa024dbb72
Last commit: 8 hours ago
Core tap JSON: 06 Mar 02:55 UTC
Core cask tap origin: https://github.com/Homebrew/homebrew-cask.git
Core cask tap HEAD: 4d46e812229bb1b3921554ee99b3f5c76720e979
Core cask tap last commit: 43 minutes ago
Core cask tap JSON: 06 Mar 02:55 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.3RFXFmajC3/org.xquartz:0
HOMEBREW_EDITOR: /usr/local/bin/bbedit
HOMEBREW_MAKE_JOBS: 8
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: octa-core 64-bit icelake
Clang: 15.0.0 build 1500
Git: 2.44.0 => /usr/local/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 13.6.4-x86_64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.2
```

### Output of `brew doctor`

```shell
Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
```

### Description of issue

Python-Mutagen has been removed from Core and the Tap is unreadable. How should I proceed?

**Comment by p-linnane** [comment, marked as answer]
Can you try `brew uninstall --force python-mutagen`?

**Reply by chrisfinazzo** [reply]
This works, but I don't know yet how this may affect the rest of my Brewfile - will continue to test.

**Reply by p-linnane** [reply]
Formulae that were using `python-mutagen` now have the PyPI resource vendored within the formulae.

**Reply by chrisfinazzo** [reply]
When creating an updated Brewfile with `bundle dump`, the `python-click` formula exhibits the same behavior - no surprise.

**Comment by thepatcher** [comment]
Is it possible to find out why this formula was removed in the first place?
And where is this kind of decision documented?

I assume this is the formula https://formulae.brew.sh/formula/python-mutagen which is not available anymore.

**Reply by ZhongRuoyu** [reply]
See Homebrew/homebrew-core#157500 for context. Documentation is in https://docs.brew.sh/Python-for-Formula-Authors, and improvements to the docs are in progress.

**Comment by ewenmcneill** [comment]
For anyone else finding this as a result of being told that `python-mutagen` or `python-click` has been removed (eg, when doing `brew update`):

```
==> Deleted Installed Formulae
python-click ✘                           python-mutagen ✘
```

the cause is the combination of the switch to vendoring these Python dependencies (as described up thread), as Formula resources rather than dependencies, *and* that the vendoring was apparently done (at least for a couple of Formula I encountered) without bumping the packaging version number (so the rebuilt versions of the reverse dependencies don't get automatically upgraded :-/ )

In my case I found the fastest way to resolve this once I'd figured out what happened was:

```
brew uninstall python-mutagen
```

to find out what my HomeBrew install thought would be broken as a result, and then:

```
brew reinstall REVERSE_DEPENDENCY
```

eg,

```
brew reinstall yt-dlp
```

to get the version built from the vendored Python packages (ie, without the dependency on the now removed `python-mutagen` Formula).  After that I could remove `python-mutagen` cleanly.

(Same for the `python-click` package, but in my case it was `pipgrip` which needed to be reinstalled with the version-not-updated-but-packaged-differently version that vendored the Python `click` library, instead of depending on it.)

For what it's worth, it'd be cleaner to have bumped the packaging version for these "package metadata changed, for the dependencies" changes, as it would give users a cleaner upgrade experience.

**ETA**: I see there was a deliberate decision to not bump the package versions of the Formula changed to vendor dependencies -- https://github.com/Homebrew/homebrew-core/issues/157500#issuecomment-1958960385 -- which seems unfortunate, given the number of packages users experienced problems with them disappearing, and hand to figure out to do a manual `reinstall` to fix.   See, eg, https://github.com/Homebrew/homebrew-core/issues/164499#issuecomment-1969449901 as another example.

Ewen
